torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -301,6 +301,89 @@ class SubgraphSlicer {
   AliasDb& aliasDb_;
   std::vector<Node*>& diff_nodes_;
 };
+
+c10::optional<bool> getProfileNodeRequiresGrad(Node* n) {
+  TORCH_INTERNAL_ASSERT(n->kind() == prim::profile);
+  if (!n->hasAttribute(attr::profiled_type)) {
+    return c10::nullopt;
+  }
+  auto& type = n->ty(attr::profiled_type);
+  if (type->castRaw<TensorType>() == nullptr) {
+    return c10::nullopt;
+  }
+  return type->expectRef<TensorType>().requiresGrad();
+}
+
+void AddRequiresGradToDifferentiableGraph(Node* diff_graph) {
+  TORCH_INTERNAL_ASSERT(diff_graph->kind() == prim::DifferentiableGraph);
+  const auto& subgraph = diff_graph->g(attr::Subgraph);
+  for (auto i : c10::irange(subgraph->outputs().size())) {
+    Value* output = subgraph->outputs()[i];
+    if (output->node()->kind() == prim::profile) {
+      // already have requires_grad info from this profile node
+      continue;
+    }
+    if (output->type()->castRaw<TensorType>() == nullptr) {
+      // non-tensors don't get profiled.
+      continue;
+    }
+    if (output->type()->expectRef<TensorType>().requiresGrad().has_value()) {
+      continue;
+    }
+
+    // this node doesn't have any requires_grad info.
+    // look at its uses to try to find a profile node.
+    c10::optional<bool> requiresGrad = c10::nullopt;
+    for (auto& use : diff_graph->output(i)->uses()) {
+      if (use.user->kind() == prim::profile) {
+        requiresGrad = getProfileNodeRequiresGrad(use.user);
+      }
+
+      // maybe the profile node got absorbed into a differentiable graph
+      if (use.user->kind() == prim::DifferentiableGraph) {
+        const auto& dg = use.user->g(attr::Subgraph);
+        // check all the uses of this graph input to look for profile nodes.
+        Value* dg_value = dg->inputs()[use.offset];
+        for (auto& dg_use : dg_value->uses()) {
+          if (dg_use.user->kind() == prim::profile) {
+            requiresGrad = getProfileNodeRequiresGrad(dg_use.user);
+            if (requiresGrad) {
+              break;
```

Comment:
nit: we are doing the early break for `if (requiresGrad)` in this loop but not the loop above